File: src/guessthedatastructure/GuessTheDataStructure.java
package guessthedatastructure;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.*;
import java.util.concurrent.ArrayBlockingQueue;

import static guessthedatastructure.DataStruct.*;
import static guessthedatastructure.Operation.*;

// https://open.kattis.com/problems/guessthedatastructure

class GuessTheDataStructure {

    public static void main(String[] args) throws FileNotFoundException {

        File file = new File("src/guessthedatastructure/in");
        Scanner scan = new Scanner(file);
        GuessTheDataStructure.getInstance();

        while (scan.hasNextLine()) {
            sb.setLength(0);
            int numOperations = Integer.parseInt(scan.nextLine());
            for (int i = 0; i < numOperations; i++) {
                sb.append(scan.nextLine()).append(" ");
            }
            DataStruct dst = calcStructure(sb);
            System.out.println(dst.name);
        }
    }

    static Stack<Integer> stack;
    static Queue<Integer> queue;
    static PriorityQueue<Integer> pq;
    static StringBuilder sb;
    static GuessTheDataStructure structs = null;

    private GuessTheDataStructure() {
        stack   = new Stack<>();
        queue   = new ArrayBlockingQueue<>(1000);
        pq      = new PriorityQueue<>(Collections.reverseOrder());
        sb      = new StringBuilder();
    }

    public static GuessTheDataStructure getInstance() {
        if (structs == null) {
            structs = new GuessTheDataStructure();
        }
        return structs;
    }

    static int[] getInts(StringBuilder line) {
        int[] cmds = new int[line.length() / 2];
        String[] strArr = line.toString().split("\\s");
        for (int i = 0; i < strArr.length; i++) {
            cmds[i] = Integer.parseInt(strArr[i]);
        }
        return cmds;
    }

    static DataStruct calcStructure(StringBuilder sb) {
        int[] cmd = getInts(sb);
        DataStruct ds = IMPOSSIBLE;

        if (stackTest(cmd)) {
            if (priorityQueueTest(cmd)) return NOT_SURE;
            if (queueTest(cmd)) return NOT_SURE;
            ds = IS_A_STACK;
        }
        if (queueTest(cmd)) {
            if (priorityQueueTest(cmd)) return NOT_SURE;
            ds = IS_A_QUEUE;
        }
        if (priorityQueueTest(cmd)) {
            if (ds == IS_A_STACK || ds == IS_A_QUEUE) return NOT_SURE;
            ds = IS_A_PRIO_QUEUE;
        }
        return ds;
    }

    // 1 2 1 1 2 1 2 2
    static boolean stackTest(int[] cmds) {

        for (int i = 0; i < cmds.length; i += 2) {
            if (cmds[i] == PUSH.id) {
                stack.push(cmds[i + 1]);
            } else if (cmds[i] == REMOVE.id) {
                if (stack.isEmpty()) return false;

                int topStack = cmds[i + 1];
                if (stack.peek() == topStack) {
                    stack.pop();
                } else {
                    stack.clear();
                    return false;
                }
            }
        }
        stack.clear();
        return true;
    }

    // 1 1 1 2 1 3 2 1 2 2 2 3
    static boolean queueTest(int[] cmds) {

        for (int i = 0; i < cmds.length; i += 2) {
            if (cmds[i] == PUSH.id) {
                queue.add(cmds[i + 1]);
            } else if (cmds[i] == REMOVE.id) {
                if (queue.isEmpty()) return false;

                int next = cmds[i + 1];
                if (queue.peek() == next) {
                    queue.remove();
                } else {
                    queue.clear();
                    return false;
                }
            }
        }
        queue.clear();
        return true;
    }

    // 1 2 1 5 1 1 1 3 2 5 1 4 2 4
    static boolean priorityQueueTest(int[] cmds) {

        for (int i = 0; i < cmds.length; i += 2) {
            if (cmds[i] == PUSH.id) {
                pq.add(cmds[i + 1]);
            } else if (cmds[i] == REMOVE.id) {
                if (pq.isEmpty()) return false;

                int next = cmds[i + 1];
                if (pq.peek() == next) {
                    pq.remove();
                } else {
                    pq.clear();
                    return false;
                }
            }
        }
        pq.clear();
        return true;
    }
}

enum DataStruct {
    IS_A_STACK("stack"), IS_A_QUEUE("queue"), IS_A_PRIO_QUEUE("priority queue"),
    NOT_SURE("not sure"), IMPOSSIBLE("impossible");

    public final String name;
    DataStruct(String name) {
        this.name = name;
    }
}

enum Operation {
    PUSH(1), REMOVE(2);

    public final int id;
    Operation(int id) {
        this.id = id;
    }
}

File: src/guessthedatastructure/go/guessthedatastructure.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

// https://open.kattis.com/problems/guessthedatastructure

const (
	PUSH = iota + 1
	POP
)

const (
	IS_STACK = uint8(iota)
	IS_QUEUE
	IS_PRIO_Q
	NOT_SURE
	IMPOSSIBLE
)

func main() {
	input := os.Getenv("GOPATH") + filepath.Join("/src/code/guessthedatastructure/in")

	file, _ := os.Open(input)
	io := bufio.NewScanner(file)

	var sb strings.Builder

	for io.Scan() {
		line := io.Text()
		cases, _ := strconv.Atoi(line)

		ops := getOperations(io, cases)
		structure := calcStructure(ops)

		// Only the 2nd test is NOT_SURE (stack + pq)
		switch structure {
		case IS_STACK:
			sb.WriteString("stack\n")
		case IS_QUEUE:
			sb.WriteString("queue\n")
		case IS_PRIO_Q:
			sb.WriteString("priority queue\n")
		case NOT_SURE:
			sb.WriteString("not sure\n")
		default:
			sb.WriteString("impossible\n")
		}
	}
	fmt.Print(sb.String())
}

func calcStructure(ops []int) uint8 {
	if stackTest(ops) {
		if queueTest(ops) || prioQTest(ops) {
			return NOT_SURE
		}
		return IS_STACK
	}
	if queueTest(ops) {
		if prioQTest(ops) {
			return NOT_SURE
		}
		return IS_QUEUE
	}
	if prioQTest(ops) {
		return IS_PRIO_Q
	}
	return IMPOSSIBLE
}

func getOperations(io *bufio.Scanner, operations int) []int {
	var cmds []int
	for i := 0; i < operations; i++ {
		io.Scan()
		line := strings.Fields(io.Text())
		num1, _ := strconv.Atoi(line[0])
		num2, _ := strconv.Atoi(line[1])
		cmds = append(cmds, num1, num2)
	}
	return cmds
}

func stackTest(ops []int) bool {
	s := &Stack{}
	for i := 0; i < len(ops); i += 2 {
		if ops[i] == PUSH {
			s.Push(ops[i+1])
		} else {
			if s.IsEmpty() {
				return false
			}
			top := ops[i+1]
			peek, _ := s.Peek()
			if peek == top {
				s.Pop()
			} else {
				return false
			}
		}
	}
	return true
}

func queueTest(ops []int) bool {
	q := &Queue{}
	for i := 0; i < len(ops); i += 2 {
		if ops[i] == PUSH {
			q.Enqueue(ops[i+1])
		} else {
			if q.IsEmpty() {
				return false
			}
			next := ops[i+1]
			peek, _ := q.Dequeue()
			if peek != next {
				return false
			}
		}
	}
	return true
}

func prioQTest(ops []int) bool {
	pq := NewMaxPQ()
	for i := 0; i < len(ops); i += 2 {
		if ops[i] == PUSH {
			pq.Insert(*NewItem(ops[i+1], ops[i+1]))
		} else {
			if pq.Len() == 0 {
				return false
			}
			next := ops[i+1]
			if pq.Extract().Priority != next {
				return false
			}
		}
	}
	return true
}
